internal/command/arguments/plan_test.go

```diff
@@ -170,6 +180,210 @@ func TestParsePlan_targets(t *testing.T) {
+				{
+					fileContent: "foo_bar.baz",
+					hasError:    false,
+				},
+				{
+					fileContent: "modu(le.boop",
+					hasError:    true,
+				},
+			},
+			want: []addrs.Targetable{foobarbaz.Subject},
+			wantDiags: tfdiags.Diagnostics(nil).Append(
+				&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Invalid syntax",
+					Detail:   `For target "modu(le.boop": Expected an attribute access or an index operator.`,
+					Subject: &hcl.Range{
+						Start: hcl.Pos{Line: 1, Column: 1},
+						End:   hcl.Pos{Line: 1, Column: 13, Byte: 12},
+					},
+				},
+			),
+		},
+		"target file valid comment": {
+			files: []testFile{
+				{
+					fileContent: "#foo_bar.baz",
+					hasError:    false,
+				},
+			},
+			want: nil,
+		},
+		"target file valid spaces": {
+			files: []testFile{
+				{
+					fileContent: "   foo_bar.baz",
+					hasError:    false,
+				},
+			},
+			want: []addrs.Targetable{foobarbaz.Subject},
+		},
+		"target file valid tab": {
+			files: []testFile{
+				{
+					fileContent: "\tfoo_bar.baz",
+					hasError:    false,
+				},
+			},
+			want: []addrs.Targetable{foobarbaz.Subject},
+		},
+		"target file valid complicated": {
+			files: []testFile{
+				{
+					fileContent: "\tmodule.boop\n#foo_bar.baz\nwow.ham",
+					hasError:    false,
+				},
+			},
+			want: []addrs.Targetable{boop.Subject, wow.Subject},
+		},
+		"target file invalid bracket with spaces": {
+			files: []testFile{
+				{
+					fileContent: `    [boop]`,
+					hasError:    true,
+				},
+			},
+			want: nil,
+			wantDiags: tfdiags.Diagnostics(nil).Append(
+				&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Invalid syntax",
+					Detail:   `For target "    [boop]": Must begin with a variable name.`,
+					Subject: &hcl.Range{
+						Start: hcl.Pos{Line: 1, Column: 1},
+						End:   hcl.Pos{Line: 1, Column: 11, Byte: 10},
+					},
+				},
+			),
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+
+			/*
+
+				1. Write temp files to storage
+					* store filenames
+				2. Create argument list from filenames, with `-target-file`
+				3. Pass argument list to ParsePlan()
+				4. Get diagnostics from ParsePlan()
+				5. Compare returned diagnostics to expected diagnostics
+					* Error if diff
+				6. Compare returned Plan to expected Plan
+					* Error if diff
+			*/
+
+			targetFileArguments := []string{}
+			wantDiagsExported := tc.wantDiags.ForRPC()
+
+			/*
+				1. For each testFile struct
+				2. Write the testFile.fileContents to a temp
+				3. Does the testfile struct have an error?
+					If yes, find the diag within the testcase that corresponds to that testfile
+			*/
+			for _, testFile := range tc.files {
+				testFile.tempFileWriter()
+				defer os.Remove(testFile.filePath)
+				targetFileArguments = append(targetFileArguments, "-target-file="+testFile.filePath)
+
+				if testFile.hasError {
+					wantDiagsExported[0].Source().Subject.Filename = testFile.filePath
```

Comment:
Note to self, assumption here that only one diag exists in the individual test. Not a great assumption, can improve this if necessary.

Reply:
Hey @apparentlymart ! This turned into quite the brainteaser! I tried out your suggestion, but it didn't get me over the line with allowing a test like this:
```
			files: []testFile{
				{fileContent: "modu(le.boop",
				{fileContent: "foo_bar.baz"},
				{fileContent: "wow^.ham",
			},
```
as the resulting `Source` duplicated the fileName across both diags.
I went down quite the journey to try and test these diags. I now think I've cracked it, but I'm again not entirely happy with the code.

Could you please have a look and let me know if I am making things too complicated [here](https://github.com/opentofu/opentofu/pull/2620/commits/ae0198e3fd7490716846804618c59579251cd1d4#diff-2d6945c1c28d40f24f2b16d9897fe68371ab7f89dd5a79ca51c5a76034f74bcaR333-R338)?